### PR TITLE
Auto-GPT V1 Async Agent

### PR DESCRIPTION
**Pull request by Swiftyos**
### Background
In order to improve the scalability and maintainability of our codebase, we have decided to implement an async agent and rearchitect the code to allow for more modular development. This decision has been based on prior discussions with the development team and aligns with our goal of making the project more flexible and adaptable to future needs.

Issue: #2458 

### Changes
- Implemented an async agent for handling asynchronous tasks and improving performance.
- Rearchitected the code base for modularity, separating concerns into individual components and making it easier to add new features or modify existing ones.
- Updated relevant internal APIs to work seamlessly with the new architecture. 
- Need to incorprate #765 

### Documentation
- Updated the in-code comments to reflect the new async agent and modular architecture.
- Updated the external documentation, including diagrams and explanations, to provide a clear understanding of the new structure and its benefits.

### Test Plan
1. Set up a local development environment with the latest changes.
2. Test the async agent by triggering multiple tasks concurrently and verifying their successful completion.
3. Test the modular components by checking their individual functionality and ensuring they work seamlessly together.
4. Conduct performance tests to ensure the new architecture maintains or improves upon the previous performance benchmarks.

### PR Quality Checklist
- [] My pull request is atomic and focuses on a single change.
- [] I have thoroughly tested my changes with multiple different prompts.
- [] I have considered potential risks and mitigations for my changes.
- [] I have documented my changes clearly and comprehensively.
- [] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by Yuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuu** [comment]
00.0

File: autogpt/agent/async_agent.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.event_queue.add_agent(self.name)
+        self.sub_mind_count = 0
+
+    async def roll_dice(self) -> int:
+        """Roll a dice and return a random number between 1 and 6.
+
+        Returns:
+            int: The result of the dice roll.
+        """
+        return random.randint(1, 4)
```
Comment (zachary-kaelan) [inline autogpt/agent/async_agent.py:79]: This is a 4-sided dice.

File: autogpt/agent/async_agent.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
+"""An asynchronous agent that can perform tasks and send and receive messages."""
+import asyncio
+import random
+from typing import List, Optional, Tuple
+
+import trio
```
Comment (Drlordbasil) [inline autogpt/agent/async_agent.py:6]: import trio_asyncio

**Comment by p-i-** [comment]
Note we have another `async` PR: https://github.com/Significant-Gravitas/Auto-GPT/pull/1245

I've put a `wontfix` label on that one, as it looks as though this is the one we want.

File: autogpt/agent/LLM/openai_provider.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    def log_debug_msg(self, message: str) -> None:
+        """Log a debug message"""
+        if self.debug_mode:
+            logger.debug(Fore.GREEN + message + Fore.RESET)
+
+    async def attempt_chat_completion(self, messages: list, attempt: int = 0) -> str:
+        """Attempt to create a chat completion"""
+
+        try:
+            return await openai.ChatCompletion.acreate(
```
Comment (p-i-) [inline autogpt/agent/LLM/openai_provider.py:65]: Has this been tested? If `.acreate` is asyncio-flavoured, we need to convert it to trio-flavoured

Throw "call an asyncio-flavoured function within trio code" into GPT4.

**Comment by ntindle** [comment]
This needs updating before merging

File: autogpt/agent/base_ai_core.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    async def run(self) -> None:
+        """The main loop of the AI core."""
+        keep_running = True
+        while keep_running:
+            await self.receive_command()
+            if self.is_sub_mind:
+                await self.receive_command_from_master()
+
+    # async def spawn_sub_mind(self, ai_core: BaseAICore) -> None:
```
Comment (lc0rp) [inline autogpt/agent/base_ai_core.py:124]: Curious to know why this is commented out.  I have been working on a crude ability to [clone the AutoGPT process ](https://github.com/KayLuke/Auto-GPT-Dolly-Plugin). This is way cleaner - it'd be a powerful feature...

Comment (Swiftyos) [reply autogpt/agent/base_ai_core.py:124]: This PR is still in development. I commented this out whilst debugging an issue. It will be coming back :)

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Pull request closed without merging**